### ISSUE TITLE
Allow exception handling to be disabled.

### ISSUE THREAD

**Issue by JodiTheTigger** [gabime/spdlog#136]
I build my code without exception handling, and spdlog cannot build as it uses exceptions.

Would you be able to make exception support optional for spdlog?

**Comment by JodiTheTigger** [comment]
cppformat uses the following method to make exceptions optional:
```cpp
#if FMT_EXCEPTIONS
# define FMT_TRY try
# define FMT_CATCH(x) catch (x)
#else
# define FMT_TRY if (true)
# define FMT_CATCH(x) if (false)
#endif

#ifndef FMT_THROW
# if FMT_EXCEPTIONS
#  define FMT_THROW(x) throw x
# else
#  define FMT_THROW(x) assert(false)
# endif
#endif
```

**Comment by gabime** [comment]
which means log errors would be just ignored if exceptions are disabled?

**Comment by JodiTheTigger** [comment]
If you follow the cppformat style, then yes. 
However instead you could have an api for error conditions that gets called instead of throwing an exception.

Either way this is just giving the choice to the user of your library to disable exceptions. It's up to them deal with the consiquences of doing so. The default would to have exceptions enabled.

**Comment by gabime** [comment]
Sorry, would cause too much code diversion.

**Comment by jgatkinsn** [comment]
This would be nice to have.  A lot of embedded systems that use C++ can't use exceptions for performance reasons.

**Comment by ConZ27** [comment]
I am in the exact same boat ... just dumped easyloggingpp because of the exceptions.

Disabling exceptions is a key requirement for our particular system and I won't be able to use this logger because of that. I do like the speed performance spdlog offers.

**Comment by gabime** [comment]
reopened due to popular request

**Comment by netheril96** [comment]
I have something similar but different: my program has exceptions enabled but I do not want a `logger` to throw any exceptions. In another words, I want a runtime option to disable exceptions so that I can drop the `try ... catch` everywhere. No one is capable of handling logging errors. What are they gonna do? Log those errors?

**Comment by tbastos** [comment]
A fairly common use case for logging is to keep track of errors in situations where you cannot throw an exception. Therefore it makes sense for a logging library to _not throw_ exceptions, or at least make it optional.
 
Moreover projects such as games are developed with -fno-exceptions and -fno-rtti, and they can't enable exceptions because of one library.

Theoretically if all `throws` in spdlog are replaced by a macro or function call, the user could redefine the macro/function to do anything, and exceptions could be made optional. When exceptions are disabled the default should be to just terminate the app. This should be easy to implement, or am I missing something?

**Comment by gabime** [comment]
Should be easy, yes. I was thinking to add to the API some kimd of set_error_handler() as suggested above by @JodiTheTigger .This way the user can terminate if he wants

**Comment by godbyk** [comment]
@gabime I'm happy to work on a PR for this as it's functionality I'd like as well. Where would be a good place to store a function pointer to a user-supplied error handler?

**Comment by gabime** [comment]
Implemented in 39cdd08a5475c63959174747a140de86c24e4849.

Exception will be thrown only if the ctor failed.

If an error happens during logging, the library will print an error message to stderr. 
To avoid flodding the screen with error messages, the rate is limited to max of 1 message/minute.

This behaviour can be changed using
```c++
spdlog::set_error_handler(new_handler_fun)
``` 
or (for specific logger) 
```c++
logger->set_error_handler(new_handler_fun):
```

**Comment by TheQuantumPhysicist** [comment]
There's a copy of fmt inside spdlog. Are the exceptions inside that disabled by default as well? So can I just mark any wrappers with `noexcept` without worrying?
